Business

sandyl289 opened the pull request:
- Wrote basic methods related to business (getAll, get by name, create)
- Changed some methods to make it similar to everyone else
- Updated methods in BusinessService to return a DTO object like everyone else
- Commented the controller methods
- I'm still writing more Business Service tests

sandyl289 committed:
Initial skeletons for DTO, Repository, Service

To do later:

BusinessRestController
- Update list of vacations

BusinessService
- deleteVacations()
- change the Check inputs to null?
- 1 updateBusiness() or 1 update method/attribute (updateName, updateAddress, updatePhoneNumber, updateEmail, updateNumberOfRepairSpots)?

sandyl289 committed:
Merge branch 'main' into business-branch

sandyl289 committed:
Update basic methods for Business Controller, Service

- Changed some methods to make it similar to everyone else
- Updated methods in BusinessService to return a DTO object like everyone else
- Commented the controller methods
- I'm still writing the Business Service tests

sandyl289 committed:
Wrote more tests for invalid emails when creating a business

JDziewonski98 commented on RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/controller/BusinessRestController.java:
```diff
@@ -0,0 +1,118 @@
+     *
+     * @return a list of all businesses
+     */
+    @GetMapping(value = {"/all", "/all/"})
```
why the value = {"/all", "/all/"}? I think @GetMapping("/all") should work, unless I misunderstand something

sandyl289 replied:
I just followed the tutorial :) I'll remove them, thanks!
EDIT: Done!

JDziewonski98 replied:
oh actually I think the original way is not so bad now that I understand it, but whatever lmao

JDziewonski98 commented on RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/controller/BusinessRestController.java:
```diff
@@ -0,0 +1,118 @@
+     * @param businessDto the update to the existing business information
+     * @return the updated business if updated successfully
+     */
+    @PostMapping(value = {"/update", "/update/"})
```
technically update requests should use PUT not POST

sandyl289 replied:
Thanks! 
@PostMapping has been changed to @PutMapping

JDziewonski98 commented on RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/controller/BusinessRestController.java:
```diff
@@ -0,0 +1,118 @@
+    public ResponseEntity<?> updateBusiness(@RequestBody BusinessDto businessDto) {
+        try {
+            return new ResponseEntity<>(businessService.updateBusiness(businessDto.getName(), businessDto.getAddress(),
+                    businessDto.getPhoneNumber(), businessDto.getEmail(), businessDto.getNumberOfRepairSpots()), HttpStatus.CREATED);
```
Careful with the HTTP status codes, instead of CREATED, I think OK makes more sense for an update

sandyl289 replied:
I didn't know "OK" existed, thanks! Change done!
For future reference: [HTTP status code definitions](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html)

JDziewonski98 commented on RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/BusinessService.java:
```diff
@@ -0,0 +1,249 @@
+        List<TimeSlot> vacations = new ArrayList<>();
+        business.setVacations(vacations);
+
+        //4) Save Business in backend
```
I appreciate the detailed comments, but I've always been told at work that only especially complicated code sections should be commented! IMO stuff like creating an arraylist, saving, and returning, etc. are self-documenting, and the comments are a little distracting. Agree / disagree?

sandyl289 replied:
Agreed. I'll try to write less comments next time

JDziewonski98 commented on RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/BusinessService.java:
```diff
@@ -0,0 +1,249 @@
+
+        //Save in BusinessRepository
+        businessRepository.save(business);
+        timeSlotRepository.save(newVacation);  //TODO: Do we need to save the new vacation to the timeslot repository?
```
i think we do save it based on the conversation we had, as it will be a 'brand new' timeslot

JDziewonski98 commented on RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/BusinessService.java:
```diff
@@ -0,0 +1,249 @@
+     */
+    @Transactional
+    public List<TimeSlot> getAllVacations() {
+        return timeSlotRepository.findAll();
```
This wont work, we have to get a specific business then call getVacations() on it

sandyl289 committed:
Remove extra values in @GetMapping

JDziewonski98 commented on RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/BusinessService.java:
```diff
@@ -0,0 +1,249 @@
+     * @author Tyler
+     */
+    private static void validateEmail(String email) throws Exception {
+        if (!matches("[A-Za-z0-9._+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}", email)) throw new Exception("Invalid email");
```
good idea, but let's put any utility methods in a new "utilities" or "helpers" package within the services package. This will be useful for all the user classes, so they can easily import it like that.

T-Syme replied:
In the ECSE 223 project, we had a util package for helper methods for converting time and date formats and stuff like that. We could make a package and start adding helper methods to it

T-Syme commented on RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/BusinessService.java:
```diff
@@ -0,0 +1,249 @@
+    public BusinessDto getBusinessByName(String name) throws Exception {
+
+        //Check input
+        if (name == null) {
```
Minor detail: should "name.equals("")" be added to this if statement too?

sandyl289 replied:
Thanks!! :) ! 
(Edit: done!)

sandyl289 committed:
Change @PostMapping to @PutMapping

- Jacek: "Update requests should use PUT not POST"

T-Syme reviewed (approved):
Apart from the Timeslot stuff that Jacek is suggesting to change, it looks good to me

sandyl289 committed:
Add extra check inputs name.equals("")

sandyl289 committed:
Change CREATED to OK

As per this suggestion (https://github.com/McGill-ECSE321-Winter2021/project-group-09/pull/75#discussion_r587935004)

sandyl289 committed:
Remove a comment

- We do need to save it in the timeSlotRepository

sandyl289 committed:
Merge branch 'main' into business-branch

sandyl289 committed:
Merge branch 'main' into business-branch

sandyl289 committed:
Add businessID attribute in BusinessDto class

sandyl289 committed:
Merge branch 'main' into business-branch

sandyl289 committed:
Transfer the validateEmail() to the utilities package

sandyl289 committed:
Add getter and setter for businessID

JDziewonski98 committed:
Merge branch 'main' into business-branch

JDziewonski98 committed:
Merge branch 'business-branch' of https://github.com/McGill-ECSE321-Winter2021/project-group-09 into business-branch

# Conflicts:
#	RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/service/utilities/ValidationHelperMethods.java
